### PR TITLE
[LoongArch] [CodeGen] Add options for Clang to generate LoongArch-specific frecipe & frsqrte instructions

### PR DESCRIPTION
**Pull request by tangaac**
Two options: `-mfrecipe` & `-mno-frecipe`.
Enable or Disable frecipe.{s/d} and frsqrte.{s/d} instructions. 
The default is `-mno-frecipe`.

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Arch/LoongArch.cpp [hunk 1]
```diff
@@ -251,6 +251,20 @@ void loongarch::getLoongArchTargetFeatures(const Driver &D,
     } else /*-mno-lasx*/
       Features.push_back("-lasx");
   }
+
+  // Select frecipe feature determined by -m[no-]frecipe.
+  if (const Arg *A =
+          Args.getLastArg(options::OPT_mfrecipe, options::OPT_mno_frecipe)) {
+    // FRECIPE depends on 64-bit FPU.
```
Comment (heiher) [inline clang/lib/Driver/ToolChains/Arch/LoongArch.cpp:258]: Why doesn't `frecipe.s` work with only `+f`?

Comment (tangaac) [reply clang/lib/Driver/ToolChains/Arch/LoongArch.cpp:258]: My Bad. I mistakenly thought that the LoongArch v1.1 instructions required fpu64 like lsx.

**Review by heiher — APPROVED** [review]
LGTM, Thanks.

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.h [hunk 2]
```diff
@@ -141,6 +141,22 @@ enum NodeType : unsigned {
   VALL_NONZERO,
   VANY_NONZERO,
 
+  // Floating point approximate reciprocal operation
+  FRECIPE_S,
```
[the remark this answers is not in the document]
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h:145]: We can't.
Each target-dependent ISD corresponds to a specific instruction in the backend.
Other instructions such as IOCSRRD,  IOCSRWR  have 4 different types of ISD: 
```c++
  IOCSRRD_B,
  IOCSRRD_W,
  IOCSRRD_H,
  IOCSRRD_D,
  IOCSRWR_B,
  IOCSRWR_H,
  IOCSRWR_W,
  IOCSRWR_D,
```

[the remark this answers is not in the document]
Comment (SixWeining) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h:145]: They are not the same. The iocsrrd_* instructions have same operand type `GPR` but frecipe_* instructions have different types: FPR32/FPR64/v8f32/...

Is it right? @wangleiat

[the remark this answers is not in the document]
Comment (wangleiat) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h:145]: > They are not the same. The iocsrrd_* instructions have same operand type `GPR` but frecipe_* instructions have different types: FPR32/FPR64/v8f32/...
> 
> Is it right? @wangleiat

yes, just like a generic ISD node, ADD/FADD ...

[the remark this answers is not in the document]
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchISelLowering.h]: The code has been updated. 
Only `FRECIPE` and  `FRSQRTE` are left.

File: clang/test/Driver/loongarch-mfrecipe.c [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+
+// CC1-FRECIPE: "-target-feature" "+frecipe"
+// CC1-NO-FRECIPE: "-target-feature" "-frecipe"
+
+// IR-FRECIPE: attributes #[[#]] ={{.*}}"target-features"="{{(.*,)?}}+frecipe{{(,.*)?}}"
+// IR-NO-FRECIPE: attributes #[[#]] ={{.*}}"target-features"="{{(.*,)?}}-frecipe{{(,.*)?}}"
+
+int foo(void) {
+    return  42;
```
Comment (SixWeining) [inline clang/test/Driver/loongarch-mfrecipe.c:29]:
```suggestion
  return 42;
```

**Review by SixWeining — APPROVED** [review]
LGTM except a nit.

**Review by SixWeining — APPROVED** [review]

**Pull request merged**